spec.bs

```diff
@@ -1068,34 +1260,79 @@ Note: Two {{PAHistogramContribution}}s for the same report can be merged if and
 
 <div algorithm>
 To <dfn>query the budget</dfn> given a [=list=] of {{PAHistogramContribution}}s
-|contributions|, an [=origin=] <var ignore>origin</var>, a [=context type=] <var
-ignore>api</var> and a [=moment=] <var ignore>currentTime</var> and a
-[=boolean=] |consumeIfPermitted|, perform the following steps. They return a
-[=list=] of [=booleans=] with the same [=list/size=] as |contributions|.
+|contributions|, a [=list=] of [=named budget ledgers=]
+|remainingBudgetLedgers|, an [=origin=] |origin|, a [=context type=] |api|, a
+[=moment=] |currentTime|, and a [=boolean=] |consumeIfPermitted|, perform the
+following steps. They return a [=list=] of [=booleans=] with the same
+[=list/size=] as |contributions|.
+
+1. If |remainingBudgetLedgers| is [=list/empty=]:
+    1. [=list/For each=] |budgetingScope| of [=all budgeting scopes=]:
+        1. Let |remainingBudgetLedger| be a new [=named budget ledger=].
+        1. [=map/Set=] |remainingBudgetLedger|[null] to |budgetingScope|'s
+            [=budgeting scope/max budget=].
+        1. [=list/Append=] |remainingBudgetLedger| to |remainingBudgetLedgers|.
+1. Let |numScopes| be the [=list/size=] of [=all budgeting scopes=].
+1. [=Assert=] that |numScopes| is equal to the [=list/size=] of
+    |remainingBudgetLedgers|.
+1. [=list/For each=] |i| in [=the exclusive range|the range=] from 0 to
+    |numScopes|, exclusive:
+    1. Let |budgetingScope| be [=all budgeting scopes=][|i|].
+    1. Let |duration| be |budgetingScope|'s [=budgeting scope/duration=].
+    1. Let |remainingBudgetLedger| be |remainingBudgetLedgers|[|i|].
+    1. [=map/For each=] |name| → |maxBudget| of |remainingBudgetLedger|:
+        1. Let |budgetConsumed| be an integer determined by an
```

Comment:
We also need to consider the case where |budgetConsumed| is small for a particular named budget, but usage of other names limits this one more than its current reservation. (This could happen if reservations aren't always consistent between calls.)

For this, I think we need an overall previous consumption or an overall budget left.